Modified vlan_pool network allocator to ensure VLAN IDs dont collide.

naved001 opened the pull request:
Issue #804 

* Raises AllocationError if an admin tries to make a network with a net_id
 from the pool that is already in use
* method validate_network_id returns a boolean if the net_id belongs to the
 allocation pool or not.
* it raises a BadArgumentError if the net_id is invalid i.e., greater than 4096
 or less than 1

I did not want to add more methods to the VlanAllocator, so decided to just modify `validate_network_id` because it was introduced only for  admin networks. Let me know if this seems like a reasonable approach to address #804. If yes, then I'll fix the tests that are broken atm and write some more tests. If not, any suggestions?

naved001 commented:
cc: @henn

zenhack commented:
I'm ok with modifying the contract for validate_network_id in theory, but the specifics of what this patch does are a little clumsy: before it just returned yes or no, is this valid? now it does a few different checks, using exceptions to report some things and the return value to report others. This is generally a sign that you're trying to do two unrelated things in one function.

I would just go ahead and add another method, `network_id_is_used` or such, that checks the new thing.

naved001 commented:
You are right; it is clumsy. I added a new commit. How does this look?

I have added two new functions; `network_id_available` and `is_network_id_in_pool`. 

I think I can merge `validate_network_id` with `network_id_available` but then if it only returns True or False then we can't differentiate between `net_id` being invalid vs not available (BadArgument vs Allocation Error). If we can move the error handling into the driver then we can raise the appropriate error and not return anything. 

`is_network_id_in_pool` would be required anyway.

naved001 commented:
@zenhack what do you think of the new commit?

zenhack commented on hil/ext/network_allocators/vlan_pool.py:
```diff
@@ -69,6 +69,20 @@ def validate_network_id(self, net_id):
+
+    def is_network_id_in_pool(self, net_id):
+        vlan = Vlan.query.filter_by(vlan_no=net_id).first()
+        return True if vlan else False
```
`vlan is not None`

zenhack commented on hil/ext/network_allocators/vlan_pool.py:
```diff
@@ -69,6 +69,20 @@ def validate_network_id(self, net_id):
         except ValueError:
             return False
 
+    def network_id_available(self, net_id):
```
This functions name suggests it merely does a check, but in fact it has a side effect. Another way to do this would be to have a `claim_network_id`, which raised an AllocationError if the id is unavailable.

naved001 replied:
> Another way to do this would be to have a claim_network_id, which raised an AllocationError if the id is unavailable.

I like this, I am gonna do that.

zenhack commented on hil/api.py:
```diff
@@ -800,7 +800,10 @@ def network_create(network, owner, access, net_id):
+        if not get_network_allocator().network_id_available(net_id):
+            raise AllocationError("Network ID is in use. Please choose a "
+                                  "different net_id")
+        allocated = get_network_allocator().is_network_id_in_pool(net_id)
```
At this point in the code, this is still always going to be False. Two sensible options:

* change this back to `allocated = False`
* move it out of the else block, so that it is executed regardless of which branch is taken, and remove `allocated = True` from the first branch.

naved001 replied:
why would it always be False? If `net_id` belongs to the vlan pool, then even though it wasn't "allocated" by HIL, we still need to mark it as `True` otherwise we can't free the network in [`network_delete`](https://github.com/CCI-MOC/hil/blob/master/hil/api.py#L838)

naved001 commented:
@zenhack I have added unit tests for vlan_pool. it's ready for review.

Izhmash commented on tests/unit/ext/network_allocators/vlan_pool.py:
```diff
@@ -76,3 +77,60 @@ def test_vlanid_for_admin_network():
+
+        api.project_create('nuggets')
+
+        # create a project owned network and get it's network_id
```
Minor, should be 'its'

naved001 replied:
changed!

zenhack commented on hil/ext/network_allocators/vlan_pool.py:
```diff
@@ -69,6 +70,21 @@ def validate_network_id(self, net_id):
+            raise BlockedError("Network ID is not available."
+                               " Please choose a different ID.")
+        else:
+            return
```
The if-else ladder here is a bit awkward. I'd do something more like:

```python

if vlan is None:
    return
elif vlan.available:
    vlan.available = False
else:
    raise BlockedError(...)
```
